### PR TITLE
operator: Add e2e tests for DSA, IAA

### LINKED ISSUE #931: e2e: add tests to deploy and test plugins with the operator
currently, our e2e tests do not test anything with the operator based flow. Add tests that install the operator, deploy plugins and tests them using the BMAAS `e2e-*` workflows.

### PR DESCRIPTION
**Pull request by ozhuraki**

### CODE REVIEW

File: test/e2e/operator/operator.go [hunk 1]
```diff
@@ -16,40 +16,116 @@
...
+		time.Sleep(timeout)
+	})
+
+	ginkgo.AfterEach(func() {
+		kubectl(ns, "delete", "-k", "deployments/operator/default/kustomization.yaml")
+
+		kubectl("node-feature-discovery", "delete", "-k", "deployments/nfd/kustomization.yaml")
+	})
+
+	ginkgo.It("deploys DSA plugin", func() {
```
Comment (mythi) [inline test/e2e/operator/operator.go:76]: FWIW, I'd be OK to replace the existing plugin deployments that use kustomize with the operator based one. After all, the kustomization .yaml's are built into the operator so they are implicitly tested.

File: pkg/controllers/iaa/controller.go [hunk 2]
```diff
@@ -124,7 +124,7 @@ func addInitContainer(ds *apps.DaemonSet, dp *devicepluginv1.IaaDevicePlugin) {
 			},
 			{
 				Name:  "IDXD_DEVICE_TYPE",
-				Value: "iaa",
+				Value: "iax",
```
Comment (mythi) [inline pkg/controllers/iaa/controller.go:127]: @ozhuraki a separate PR for this?

File: test/e2e/operator/operator.go (hunk 1 above)
Comment (mythi) [reply test/e2e/operator/operator.go:76]: This would also be closer to what we agreed earlier: have these operator tests in their own device specific containers (ginkgo term) instead of in a single operator one.

File: test/e2e/dsa/dsa.go [hunk 3]
```diff
@@ -42,6 +58,24 @@ func describe() {
 	f := framework.NewDefaultFramework("dsaplugin")
 	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
 
+	ginkgo.BeforeEach(func() {
+		kubectl("node-feature-discovery", "apply", "-k", "deployments/nfd/kustomization.yaml")
```
Comment (mythi) [inline test/e2e/dsa/dsa.go:62]: as commented in https://github.com/intel/intel-device-plugins-for-kubernetes/issues/1143#issuecomment-1277547731 it would probably be good to deploy NFD only once for all

Comment (hj-johannes-lee) [reply test/e2e/dsa/dsa.go:62]: Is there any difference in this pr actually?
Only one It() exists.
Is there a case that we run all tests at once?

File: Makefile [hunk 4]
```diff
@@ -157,10 +157,10 @@ e2e-gpu:
 	@$(GO) test -v ./test/e2e/... -ginkgo.v -ginkgo.progress -ginkgo.focus "GPU" -delete-namespace-on-failure=false
 
 e2e-dsa:
-	@$(GO) test -v ./test/e2e/... -ginkgo.v -ginkgo.progress -ginkgo.focus "DSA" -delete-namespace-on-failure=false
+	@$(GO) test -v ./test/e2e/... -ginkgo.v -ginkgo.progress -ginkgo.focus "DSA" -delete-namespace-on-failure=false -timeout=2400s -ginkgo.timeout=2400s
 
 e2e-iaa:
-	@$(GO) test -v ./test/e2e/... -ginkgo.v -ginkgo.progress -ginkgo.focus "IAA" -delete-namespace-on-failure=false
+	@$(GO) test -v ./test/e2e/... -ginkgo.v -ginkgo.progress -ginkgo.focus "IAA" -delete-namespace-on-failure=false -timeout=2400s -ginkgo.timeout=2400s
```
Comment (bart0sh) [inline Makefile:163]: Does this mean we're increasing test time to 40 min?

File: test/e2e/deviceplugins_suite_test.go [hunk 5]
```diff
@@ -32,21 +33,36 @@ import (
...
+	"github.com/intel/intel-device-plugins-for-kubernetes/test/e2e/utils"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/component-base/logs"
 	"k8s.io/component-base/version"
 	"k8s.io/klog/v2"
 	"k8s.io/kubernetes/test/e2e/framework"
 	"k8s.io/kubernetes/test/e2e/framework/config"
-	"k8s.io/kubernetes/test/e2e/framework/kubectl"
+	k "k8s.io/kubernetes/test/e2e/framework/kubectl"
```
Comment (bart0sh) [inline test/e2e/deviceplugins_suite_test.go:44]: not sure it's a good idea to use 'k' instead of 'kubelet'. Later is more explicit and readable in the code.

File: test/e2e/dsa/dsa.go [hunk 6]
```diff
@@ -22,18 +22,34 @@ import (
...
 const (
+	ns                = "inteldeviceplugins-system"
+	timeout           = time.Second * 180
 	kustomizationYaml = "deployments/dsa_plugin/overlays/dsa_initcontainer/dsa_initcontainer.yaml"
 	configmapYaml     = "demo/dsa.conf"
 	demoYaml          = "demo/dsa-accel-config-demo-pod.yaml"
 	podName           = "dsa-accel-config-demo"
 )
 
+func kubectl(ns string, cmd string, opt string, file string) {
```
Comment (bart0sh) [inline test/e2e/dsa/dsa.go:39]: This funciton is duplicated in dsa and iaa.

File: deployments/operator/manager/manager.yaml [hunk 7]
```diff
@@ -26,13 +26,6 @@ spec:
       - image: intel/intel-deviceplugin-operator:devel
         imagePullPolicy: IfNotPresent
         name: manager
-        resources:
```
Comment (mythi) [inline deployments/operator/manager/manager.yaml:29]: `manager.yaml` changes need to be in its own PR with a commit message *why* this it is justified. I don't think it's the right fix that we drop these completely.

File: Makefile (hunk 4 above)
Comment (ozhuraki) [reply Makefile:163]: @bart0sh 
It's a good question. In a wost case, yes.

I am not entirely sure, this is needed and will check if we could drop it (there's some problem within e2e framework ocassionaly, haven't managed to get to the root cause for it yet)

File: test/e2e/dsa/dsa.go (hunk 3 above)
Comment (ozhuraki) [reply test/e2e/dsa/dsa.go]: @hj-johannes-lee 
> Is there a case that we run all tests at once?

We actually run them all at once

> Ah, well, maybe we do not need to copy and paste for all device plugins?

Solved, please take a look

Comment (hj-johannes-lee) [reply test/e2e/dsa/dsa.go]: @ozhuraki I mean, do we run just "go test -v ./test/e2e/..." without "-args -ginkgo.focus <something>"? Is there a system that can test all plugins e2e tests? Sorry, maybe I was not clear.

But, anyway you changed since you saw it as better, so I believe it is better.! :)

File: test/e2e/dsa/dsa.go [hunk 8]
```diff
@@ -96,4 +98,26 @@ func describe() {
 
 		framework.Logf("log output: %s", log)
 	})
+
+	ginkgo.It("deploys DSA plugin with operator", func() {
```
Comment (mythi) [inline test/e2e/dsa/dsa.go:102]: I wonder where's the code path the undeploys the other daemonset before running this (or vice versa)?

**Comment by mythi** [comment]
> `Internal error occurred: failed calling webhook "webhook.cert-manager.io": failed to call webhook: Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": context deadline exceeded`

@ozhuraki I've seen this error for the past ~2 rounds of your pushesh. Interestingly, e2e-iaa is OK... something with the operator manager deployment seems to trigger these. Just making wild guesses but long time ago setting `GOMAXPROCS=1` env variable to manager helped with some cpu throttling problems.

**Comment by ozhuraki** [comment]
Changes:
- Added a check for NFD readiness
- Removed CPU limits for the operator under e2e
- Increased DSA, IAA webhook timeouts
- Added GOMAXPROCS to the operator e2e deployment
- Dropped BeforeEeach, AfterEeach in SGX (SGX's AfterEach deleted NFD).
  After the NFD was moved to BeforeSuite, this started to interfere,
  yet to occasionally producing an OK overal pass

File: test/e2e/dsa/dsa.go (hunk 8 above)
Comment (ozhuraki) [reply test/e2e/dsa/dsa.go:102]: @mythi 
k8s e2e deletes a namespace after the test

**Comment by mythi** [comment]
> Changes:
> 
>     * Added a check for NFD readiness

The NFD rework can go to a PR of its own. The same for the `initImage` additions to DSA/IAA resource YAMLs.

> 
>     * Removed CPU limits for the operator under e2e
> 
>     * Increased DSA, IAA webhook timeouts
> 
>     * Added GOMAXPROCS to the operator e2e deployment

Do we understand the problem and *why* these (and the unconditional 2 minutes sleep) are needed?

> 
>     * Dropped BeforeEeach, AfterEeach in SGX (SGX's AfterEach deleted NFD).
>       After the NFD was moved to BeforeSuite, this started to interfere,
>       yet to occasionally producing an OK overal pass

Yes this is the right thing to do. I wonder if we have similar race conditions elsewhere...

File: test/e2e/dsa/dsa.go [hunk 9]
```diff
@@ -96,4 +98,28 @@ func describe() {
...
+		utils.Kubectl(ns, "apply", "-k", "deployments/operator/e2e/kustomization.yaml")
+
+		if _, err := e2epod.WaitForPodsWithLabelRunningReady(f.ClientSet, ns, labels.Set{"control-plane": "controller-manager"}.AsSelector(), 1, timeout); err != nil {
+			framework.Failf("unable to wait for all pods to be running and ready: %v", err)
+		}
+
+		time.Sleep(timeout)
```
Comment (bart0sh) [inline test/e2e/dsa/dsa.go:109]: what we're waiting for here?

File: test/e2e/dsa/dsa.go (hunk 8 above)
Comment (bart0sh) [reply test/e2e/dsa/dsa.go:102]: Yes, but it happens *after* the test, right? As far as I understood, @mythi asked about un-deploying whatever is deployed previously in this test.

**Review by mythi — COMMENTED** [review]
a couple of comments before the rebase.

**Comment by ozhuraki** [comment]
Changes:
- Dropped a timeout after the operator deployment
- Added a check for the webhook availability
  Apparently a simple curl check (--connect-timeout ...) isn't enough.
  Despite the check succeeding, webhook occasionally timeouts at CR creation
  (even with e2e customizations: increased timeout, no CPU limit).
- Dropped operator e2e customizations

File: test/e2e/dsa/dsa.go [hunk 10]
```diff
@@ -96,4 +98,28 @@ func describe() {
...
+			framework.Failf("unable to wait for all pods to be running and ready: %v", err)
+		}
+
+		if err := utils.WaitForNodesWithResource(f.ClientSet, "dsa.intel.com/wq-user-dedicated", timeout); err != nil {
+			framework.Failf("unable to wait for nodes to have positive allocatable resource: %v", err)
+		}
+
+		utils.Kubectl(ns, "delete", "-f", "deployments/operator/samples/deviceplugin_v1_dsadeviceplugin.yaml")
+
+		f.DeleteNamespace(ns)
```
[the remark this answers is not in the document]
Comment (ozhuraki) [reply test/e2e/dsa/dsa.go:125]: >Maybe that's causing issues?
Theoretically yes, however not in this case, i.e. we have a clean run for each worklflow.

Everything succeeds, so namespace isn't a problem here too ("ns" is propagated to RunKubectlOrDie()).

An unreliable webhook availability in a CPU constrained environent is a problem.

Ther previous problems were caused by SGX's AfterEach which is fixed.

[the remark this answers is not in the document]
Comment (mythi) [reply test/e2e/dsa/dsa.go:125]: You could try if something like this would help instead:
```
-	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+	if err := mgr.AddReadyzCheck("readyz", mgr.GetWebhookServer().StartedChecker()); err != nil {
```

**Comment by ozhuraki** [comment]
Changes:
- Changed the order of applied customizations in the default operator deployment
  preventing a clean undeployment.
  Apparently this is a long standing bug leading to sporadic errors
  on operator undeployment (even in the simplest scenario: manual operator deploy/undeploy)
- Added the deletion of deployed operator resources
- Increased the terminationGracePeriodSeconds.
  The default also results in errors on undeployment
- Dropped the namespace from the deletion kubectl (deployments/operator/default)
  This also results in error under k8s/e2e

**Comment by mythi** [comment]
> * Changed the order of applied customizations in the default operator deployment
>       preventing a clean undeployment.
>       Apparently this is a long standing bug leading to sporadic errors
>       on operator undeployment (even in the simplest scenario: manual operator deploy/undeploy)

kustomize orders its output. What is the error you are seeing? I tried `kubectl kustomize` with and without this change and the output yaml is the same.

File: test/e2e/dsa/dsa.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mythi) [reply test/e2e/dsa/dsa.go:127]: For the record, `utils.TestWebhookServerTLS()` is not going to be acceptable.

**Comment by ozhuraki** [comment]
Changes:
- Dropped the changes in the order of the operator customizations

**Comment by ozhuraki** [comment]
Changes:
- Dropped a namespace at CR creation / deletion

File: test/e2e/dsa/dsa.go [hunk 11]
```diff
@@ -96,4 +98,28 @@ func describe() {
...
+
+		if err := utils.WaitForNodesWithResource(f.ClientSet, "dsa.intel.com/wq-user-dedicated", timeout); err != nil {
+			framework.Failf("unable to wait for nodes to have positive allocatable resource: %v", err)
+		}
+
+		utils.Kubectl("", "delete", "-f", "deployments/operator/samples/deviceplugin_v1_dsadeviceplugin.yaml")
+
+		utils.Kubectl("", "delete", "-k", "deployments/operator/default/kustomization.yaml")
+
+		f.DeleteNamespace(ns)
```
Comment (mythi) [inline test/e2e/dsa/dsa.go:123]: this can be dropped now, right?

File: test/e2e/iaa/iaa.go [hunk 12]
```diff
@@ -96,4 +98,28 @@ func describe() {
...
+
+		if err := utils.WaitForNodesWithResource(f.ClientSet, "iaa.intel.com/wq-user-dedicated", timeout); err != nil {
+			framework.Failf("unable to wait for nodes to have positive allocatable resource: %v", err)
+		}
+
+		utils.Kubectl("", "delete", "-f", "deployments/operator/samples/deviceplugin_v1_iaadeviceplugin.yaml")
+
+		utils.Kubectl("", "delete", "-k", "deployments/operator/default/kustomization.yaml")
+
+		f.DeleteNamespace(ns)
```
Comment (mythi) [inline test/e2e/iaa/iaa.go:123]: drop?

File: cmd/operator/main.go [hunk 13]
```diff
@@ -194,7 +194,7 @@ func main() {
 		}
 	}
 
-	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+	if err := mgr.AddReadyzCheck("readyz", mgr.GetWebhookServer().StartedChecker()); err != nil {
```
Comment (mythi) [inline cmd/operator/main.go:197]: with this we have two `readyz` probes. Is that expected?

**Comment by ozhuraki** [comment]
Changes:
- Dropped a namespace when deploying the operator
- Removed the deletion of the namespace
- Corrected a previous mistake with AddReadyzCheck()

**Review by mythi — APPROVED** [review]
@ozhuraki thanks! we're finally past the simics woes and the end result looks good
